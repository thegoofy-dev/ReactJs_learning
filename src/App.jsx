import { useEffect, useState } from "react";
import { PostContainer } from "./Components/POST/PostContainer";
import { UserContext } from "./utils/contexts/UserContext";
import { useFetchUser } from "./utils/hooks/useFetchUser";
import { Outlet, Link, useNavigate } from "react-router-dom";

export default function App() {
  const { user, error, loading } = useFetchUser(1);
  // console.log("User:", user);
  // console.log("Loading:", loading);
  // console.log("Error:", error);

  const [userData, setUserData] = useState();

  const navigate = useNavigate();

  useEffect(() => {
    if (!loading && !error && user) setUserData(user);
  }, [loading, user, error]);

  return (
    <>
      <nav>
        <ul>
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
            <Link to="/users">Users</Link>
          </li>
          <li>
            <Link to="/blog-post">Blogs</Link>
          </li>
        </ul>
      </nav>

      <div>
        <label htmlFor="data">Enter Data:</label>
        <input type="text" id="data" onChange={(e) => {
          console.log(e.target.value);
          if(e.target.value.length > 10)
            navigate("/blog-post");
        }} /> 
      </div>

      <UserContext.Provider value={{ ...userData, setData: setUserData }}>
        <div>{loading ? "Loading..." : <PostContainer />}</div>
      </UserContext.Provider>
      <Outlet />
    </>
  );
}
