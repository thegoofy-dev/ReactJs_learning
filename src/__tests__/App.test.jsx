import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";
import { within } from "@testing-library/react";

describe("App", () => {
  it("should render users", () => {
    const { container } = render(<App usersData={[]} />);
    expect(container).toMatchSnapshot();
  });
});

describe("When there is only 1 user", () => {
  describe("Edit Button is Clicked", () => {
    it("should render save button", async () => {
      render(
        <App
          usersData={[
            {
              id: 1,
              username: "John",
              email: "john@gmail.com",
            },
          ]}
        />
      );
      const editButton = screen.getByRole("button", { name: "Edit" });
      await userEvent.click(editButton);
      const saveButton = screen.getByRole("button", { name: "Save" });
      expect(saveButton).toBeInTheDocument();
    });

    it("should display username & email input fields", async () => {
      render(
        <App
          usersData={[
            {
              id: 1,
              username: "John",
              email: "john@gmail.com",
            },
          ]}
        />
      );
      const editButton = screen.getByRole("button", { name: "Edit" });
      await userEvent.click(editButton);
      expect(screen.getByLabelText("Username:")).toBeInTheDocument();
      expect(screen.getByLabelText("Email:")).toBeInTheDocument();
    });
  });
});

describe("When there are 2 users", () => {
  it("should have 2 users", () => {
    render(
      <App
        usersData={[
          {
            id: 1,
            username: "John",
            email: "john@gmail.com",
          },
          {
            id: 2,
            username: "Michael",
            email: "michael@gmail.com",
          },
        ]}
      />
    );

    expect(screen.getByText("John")).toBeInTheDocument();
    expect(screen.getByText("Michael")).toBeInTheDocument();
  });

  it("should display save button for 1st user when edit button is clicked", async () => {
    render(
      <App
        usersData={[
          {
            id: 1,
            username: "John",
            email: "john@gmail.com",
          },
          {
            id: 2,
            username: "Michael",
            email: "michael@gmail.com",
          },
        ]}
      />
    );
    const userDetails = screen.getByTestId("user-details-1");
    expect(within(userDetails).queryByText("Michael")).toBeNull();
    const editBtn = within(userDetails).getByRole("button", { name: "Edit" });
    await userEvent.click(editBtn);
    expect(
      within(userDetails).getByRole("button", { name: "Save" })
    ).toBeInTheDocument();
  });
});
