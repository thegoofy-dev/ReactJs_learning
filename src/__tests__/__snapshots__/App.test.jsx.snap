// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render users 1`] = `
<div>
  <div>
    <b>
      ID:
    </b>
     
    <span>
      1
    </span>
     
    <br />
    <b>
      Username: 
    </b>
    <span>
      John
    </span>
    <br />
    <b>
      Email: 
    </b>
    <span>
      john@gmail.com
    </span>
    <br />
    <div>
      <button
        data-testid="edit-btn-1"
      >
        Edit
      </button>
      <button
        data-testid="delete-btn-1"
      >
        Delete
      </button>
    </div>
    <hr />
  </div>
  <div>
    <b>
      ID:
    </b>
     
    <span>
      2
    </span>
     
    <br />
    <b>
      Username: 
    </b>
    <span>
      Michael
    </span>
    <br />
    <b>
      Email: 
    </b>
    <span>
      michael@gmail.com
    </span>
    <br />
    <div>
      <button
        data-testid="edit-btn-2"
      >
        Edit
      </button>
      <button
        data-testid="delete-btn-2"
      >
        Delete
      </button>
    </div>
    <hr />
  </div>
</div>
`;
